src/frontend/SignatureMismatch.ml

```diff
@@ -153,6 +166,16 @@ let extract_function_types f =
       Some (return, args, (fun x -> UserDefined x), mem)
   | _ -> None
 
+let promote es promotions =
+  List.map2_exn es promotions ~f:(fun (exp : Ast.typed_expression) prom ->
+      let emeta = exp.emeta in
+      match prom with
+      | RealPromotion when emeta.type_ <> UReal ->
+          Ast.{expr= Ast.Promotion (exp, UReal); emeta}
```

Comment:
You check for `UReal` but what if `emeta.type_` is `UArray UReal`?
And shouldn't the expression type actually change, update `emeta`?